File: Arrays/02-second_smallest/Python/brute.py
from typing import List

# Brute Force Approach
# ---------------------
# INTUITION - Sort the array then give the second element if any
# TIME COMPLEXITY - O(N LOGN) - due to sorting
# SPACE COMPLEXITY - O(N) - to store the sorted array

def find_second_smallest_brute( array : List[int] ) -> int :
    sorted_array = sorted(set(array), reverse=True)
    return sorted_array[-2] if len(sorted_array) > 1 else None


# CHECK
arr = [2, 1, 5, 9, 100]
print(f"Second smallest in {arr} : {find_second_smallest_brute(arr)}")

File: Arrays/02-second_smallest/C++/brute.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;

/*
Brute Force Approach
---------------------
INTUITION - Sort the array then give the second last element if any
TIME COMPLEXITY - O(N LOGN) - due to sorting
SPACE COMPLEXITY - O(N) - to store the sorted array
*/

int find_second_smallest_brute( vector<int> array ){
    sort(array.begin(), array.end());
    return array.size() > 1 ? array[array.size()-2] : NULL;
}

int main() {
    vector<int> vec = {5};
    cout << "Second smallest : " << find_second_smallest_brute(vec) << endl;
    return 0;

}